compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp

```diff
@@ -450,220 +451,245 @@ void LLVMTarget::populateDefaultsFromTargetMachine() {
+#define LLVM_INITIALIZE_TARGET_VE()
+#define LLVM_INITIALIZE_TARGET_XCore()
+
+#define LLVM_TARGET(TargetName) LLVM_INITIALIZE_TARGET_##TargetName()
+#include "llvm/Config/Targets.def"
 }
 
-// static
-void LLVMTargetOptions::initializeFromFlags(LLVMTargetOptions &targetOptions) {
-  targetOptions.initializeTargetInvariantFlags();
-
-  // Target parameters.
-  static llvm::cl::opt<std::string> clTargetTriple(
-      "iree-llvmcpu-target-triple",
-      llvm::cl::desc("LLVM target machine triple."),
-      llvm::cl::init(llvm::sys::getProcessTriple()));
-  static llvm::cl::opt<std::string> clTargetCPU(
-      "iree-llvmcpu-target-cpu",
-      llvm::cl::desc(
-          "LLVM target machine CPU; use 'host' for your host native CPU."),
-      llvm::cl::init("generic"));
-  static llvm::cl::opt<std::string> clTargetCPUFeatures(
-      "iree-llvmcpu-target-cpu-features",
-      llvm::cl::desc("LLVM target machine CPU features; use 'host' for your "
-                     "host native CPU."),
-      llvm::cl::init(""));
-  static llvm::cl::opt<bool> clLinkEmbedded(
-      "iree-llvmcpu-link-embedded",
-      llvm::cl::desc("Links binaries into a platform-agnostic ELF to be loaded "
-                     "by the embedded IREE ELF loader."),
-      llvm::cl::init(LLVMTarget::DEFAULT_LINK_EMBEDDED));
-  std::optional<LLVMTarget> maybeTarget =
-      LLVMTarget::create(clTargetTriple, clTargetCPU, clTargetCPUFeatures,
-                         /*requestLinkEmbedded=*/clLinkEmbedded);
-  if (maybeTarget) {
-    targetOptions.target = *maybeTarget;
-  } else {
-    llvm::errs() << "Inconsistency in iree-llvmcpu-target-cpu-* command-line "
-                    "flags. The target CPU is not properly defined.\n";
-  }
-  LLVMTarget &target = targetOptions.target;
+//===----------------------------------------------------------------------===//
+//    __    __   ___________    ____     ____    ____  ______    __    __     //
+//   |  |  |  | |   ____\   \  /   /     \   \  /   / /  __  \  |  |  |  |    //
+//   |  |__|  | |  |__   \   \/   /       \   \/   / |  |  |  | |  |  |  |    //
+//   |   __   | |   __|   \_    _/         \_    _/  |  |  |  | |  |  |  |    //
+//   |  |  |  | |  |____    |  |   __        |  |    |  `--'  | |  `--'  |    //
+//   |__|  |__| |_______|   |__|  (_ )       |__|     \______/   \______/     //
+//                                 |/                                         //
+//===----------------------------------------------------------------------===//
+//
+// Beware adding command-line flags here: IREE is a cross-compiler and can
+// compile for multiple targets in a single invocation. Global flags added here
+// apply to all targets with no way to override them from hosting applications
+// that may need to programmatically set them per target and that's bad.
+//
+// Flags *must* be added to the LLVMTarget if they are target-specific and
+// LLVMTargetOptions if they are apply to the whole backend.
```

Comment:
FYI, I moved this comment here and reworded it slightly, since this is the only place where plugin flags can be defined (or... maybe they can be defined in other places, but it won't do what you want? idk :P)

Reply:
its weird we had flags not in a category here/global. It looks like you cleaned up most of that though.

And in future for (say PJRT) most of these would actually be on the Module rather?

Reply:
Yeah, Ben has been cleaning up the flags as part of multi-device/multi-targeting work.

In the case of compiling for multiple CPU targets (or multiple variants of the same target) in a single .vmfb, the existing flags are inflexible. What you really want is a list of device configurations, with each configuration including details about tile sizes, available CPU features, etc. Those configurations can be...
* embedded into the compiler itself (like target triple -> groups of settings)
* embedded into .mlir files, e.g. when coming directly from frontends or built into a JIT
* specified in sidecar files (`--iree-hal-device-config=config_x86.mlir --iree-hal-device-config=config_arm.mlir`)

That's my view of where things are headed anyways. Ben also wrote up some plans [here on Discord](https://discord.com/channels/689900678990135345/689906000043573354/1214738272459423745)

For this code in particular, there are some target-invariant options like linker paths and then _lots_ of target-specific options that the CLI flags will only be used to populate default values for. In the multitargeting case, all of those will come from configs.